### PR TITLE
Improve documentation on contributing to Ruby

### PR DESCRIPTION
**Pull request by jemmaissroff**
co-authored-by: Peter Zhu <peter@peterzhu.ca>
co-authored-by: Stan Lo <stan001212@gmail.com>

### CODE REVIEW

File: doc/contributing/building_ruby.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+2. Install optional, recommended dependencies:
+
+    * OpenSSL/LibreSSL
+    * readline/editline(libedit
+    * zlib
+    * libffi
+    * libyaml
+    * libexecinfo (FreeBSD)
+    * NDBM/QDBM
```
Comment (jeremyevans) [inline doc/contributing/building_ruby.md:21]: The dbm and gdbm extensions were removed, so the last two bullets can be removed

File: doc/contributing/building_ruby.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    * C compiler
+    * autoconf - 2.67 or later
+    * bison - 2.0 or later
+    * gperf - 3.0.3 or later
+    * ruby - 2.7 or later
+
+2. Install optional, recommended dependencies:
+
+    * OpenSSL/LibreSSL
+    * readline/editline(libedit
```
Comment (jeremyevans) [inline doc/contributing/building_ruby.md:16]: Missing `)` here

File: doc/contributing/making_changes_to_ruby.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+## Code style
+
+Here are some general rules to follow when writing Ruby and C code for CRuby:
+
+* Do not change code unrelated to your pull request (including style fixes)
+* Indent 4 spaces for C without tabs (tabs are two levels of indentation, equivalent to 8 spaces)
+* Indent 2 spaces for Ruby without tabs
+* ANSI C style for 1.9+ for function declarations
+* Follow C90 (not C99) Standard
```
Comment (jeremyevans) [inline doc/contributing/making_changes_to_ruby.md:21]: C99 is now allowed since Ruby 2.7.

File: doc/contributing/making_changes_to_ruby.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+# Contributing a pull request
+
+## Code style
+
+Here are some general rules to follow when writing Ruby and C code for CRuby:
+
+* Do not change code unrelated to your pull request (including style fixes)
+* Indent 4 spaces for C without tabs (tabs are two levels of indentation, equivalent to 8 spaces)
+* Indent 2 spaces for Ruby without tabs
+* ANSI C style for 1.9+ for function declarations
```
Comment (jeremyevans) [inline doc/contributing/making_changes_to_ruby.md:20]: Probably can drop the `for 1.9+` part

File: doc/contributing/building_ruby.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    * libffi
+    * libyaml
+    * libexecinfo (FreeBSD)
+    * NDBM/QDBM
+    * GDBM
+
+3. Checkout the CRuby source code:
+
+    ```
+    git clone https://github.com/ruby/ruby.git ruby-master
```
Comment (junaruga) [inline doc/contributing/building_ruby.md:27]: What's the point to use the specific directory `ruby-master` for the git repository? The directory is not only for master branch, but also all the branches. I think just `git clone https://github.com/ruby/ruby.git` might be better.

For example, in the following case, the `ruby-master` is not for the master branch any more.

```
$ git clone https://github.com/ruby/ruby.git ruby-master
$ cd ruby-master
$ git checkout v3_1_2
```

Comment (duerst) [reply doc/contributing/building_ruby.md:27]: > What's the point to use the specific directory `ruby-master` for the git repository? The directory is not only for master branch, but also all the branches. I think just `git clone https://github.com/ruby/ruby.git` might be better.

Yes. If somebody has several checkouts, they will probably have their own system of naming them. And if we every decide to change the name of the main branch from 'master' to something else, this makes it one less place we'd have to change.

**Comment by hsbt** [comment]
`CONTRIBUTING.md` is the special feature of GitHub. I'm not sure if this change is working with the same behavior.

FYI: https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors#adding-a-contributing-file

**Comment by hsbt** [comment]
I'm okay to rename `doc` to `docs`.

**Comment by st0012** [comment]
@hsbt We decided to add the `CONTRIBUTING.md` back so the renaming is not necessary now 😄

**Comment by ioquatix** [comment]
My hacking guide was deleted and it has information that is not reproduced in any other documentation... :(

**Comment by mame** [comment]
@jemmaissroff Why did you delete "doc/make_cheatsheet.md"? 😕 It included how to run a test individually, which was very useful tips for debugging. How to measure coverage is also important for me.

**Comment by ioquatix** [comment]
I'm planning to restore hacking guide.. @mame maybe we can collaborate on "non-mainstream" documentation but is useful for developers of Ruby. I was going to add `doc/hacking/...`

**Comment by st0012** [comment]
@mame @ioquatix Sorry about that 🙏  We didn't know they're still in used and thought we could just extract the gist of them to the new building/testing Ruby guide.

I now see `hacking.md` has been restored, so we can open another PR to restore `make_cheatsheet.md` too.

I also did a quick search and these seem to be the only remaining docs that talk about advanced `make` techniques. Do you think we can merge them together and put it under `contributing` as well? It may look like `doc/contributing/advanced_techniques.md`.

**Comment by ioquatix** [comment]
I'm fine with making a cohesive set of documentation.

**Comment by jemmaissroff** [comment]
@mame @ioquatix - I definitely want to echo @st0012's apologies for this. The goal of this PR was to reduce duplication (/ different information on the same topic in different places), and streamline the docs into one cohesive set of documentation. We apologize for the documentation that we dropped when making this attempt.

My concern with the full introduction of the hacking guide is that we've now again duplicated documentation that this PR intended to unify. For instance, the ["Configure Ruby"](https://github.com/ruby/ruby/blob/master/doc/hacking.md#configure-ruby) section of the hacking guide is partially duplicated in [Building Ruby](https://github.com/ruby/ruby/blob/master/doc/contributing/building_ruby.md#quick-start-guide), and the ["Running tests" section](https://github.com/ruby/ruby/blob/master/doc/hacking.md#running-tests) is duplicated in [Testing Ruby](https://github.com/ruby/ruby/blob/master/doc/contributing/testing_ruby.md).

I see that there are certain sections, and parts of sections, we missed when trying to consolidate this documentation. My strong preference would be to avoid re-introducing duplication. Duplication in documentation, especially partial duplication, can oftentimes leave whoever is trying to follow the documentation confused about what is out of date, why it's out of sync, and where the source of truth exists.

The clearest way I see to remove duplication is to migrate the sections we missed from `hacking.md`, along with the sections we missed from `doc/make_cheatsheet.md`,  into the new consolidated format. If this is acceptable to you, I will gladly make a PR to do this and send it to both of you. Please again accept my apologies for removing documentation that was in use, that was never the intention here.

**Comment by ioquatix** [comment]
I'm fine with merging the docs provided I have a place to add notes. I'm glad we are improving the documentation.

**Comment by jemmaissroff** [comment]
@ioquatix @mame @st0012 - I made #5963 in an attempt to restore the information from `doc/make_cheatsheet.md` and `doc/hacking.md` into the new cohesive format. Please take a look at your convenience, and let me know if it still seems there is anything missing. Again, the biggest goals here are de-duplication and a clear structure for documentation.

**Pull request merged**